### PR TITLE
Add Graph to Spaces

### PR DESCRIPTION
**Pull request by jjshoots**
# Description

This adds graph spaces to Gym.

Essentially, this PR adds support of Gym spaces to also include graph representations.
Graphs are represented using a `namedtuple`.
Graph spaces by default accept either `Discrete` or `Box` spaces as `node` or `edge` features, while the `node_list` is an `n x 2` integer only vector representing the list of nodes that each edge is connected to.

The space does not make any restrictions on amount of edges between nodes, nor does it restrict graphs from having self-connections, and further does not make any assumptions or restrictions on whether edges are unidirectional or bidirectional. 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: gym/spaces/graph.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+class GraphObj:
+    r"""A base for constructing information as graphs."""
+
+    def __init__(
+        self,
+        nodes: np.ndarray,
+        edges: Optional[np.ndarray] = None,
+        edge_links: Optional[np.ndarray] = None,
+    ):
+        r"""Constructor for Graph information.
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:22]: Could you use the google docstring style?

File: gym/spaces/graph.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+from gym.spaces.box import Box
+from gym.spaces.discrete import Discrete
+from gym.spaces.multi_discrete import MultiDiscrete
+from gym.spaces.space import Space
+from gym.utils import seeding
+
+
+class GraphObj:
+    r"""A base for constructing information as graphs."""
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:14]: Could you improve this docstring with example code

File: gym/spaces/graph.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+        The argument ``edge_space`` specifies the base space that each edge feature will use.
+        This argument must be either a Box or Discrete instance.
+
+        Args:
+            node_space (Union[Box, Discrete]): space of the node features.
+            edge_space (Union[None, Box, Discrete]): space of the node features.
+            seed: Optionally, you can use this argument to seed the RNG that is used to sample from the space.
+        """
+        self._np_random = None
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:106]: Why is this necessary, should this not be passed to the `super().__init__` which will handle this?

File: gym/spaces/graph.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+        self.node_space = node_space
+        self.edge_space = edge_space
+
+        # graph object creator
+        self.graphObj = GraphObj
+
+        super().__init__(None, None, seed)
+
+    def _generate_sample_space(self, base_space, num) -> Optional[Union[Box, Discrete]]:
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:129]: Add `base_space` and `num` type?

File: gym/spaces/graph.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        sampled_edges = self._sample_sample_space(edge_sample_space)
+
+        sampled_node_list = None
+        if sampled_edges is not None:
+            if num_edges > 0:
+                sampled_node_list = self.np_random.integers(
+                    low=0, high=num_edges, size=(num_edges, 2)
+                )
+
+        return GraphObj(sampled_nodes, sampled_edges, sampled_node_list)
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:180]: Is this graph object necessary? Could we not return a tuple of nodes, adjacency matrix and sampled node list. 
Here, `sampled_node_list` and in graphobj `edge_links` have completely different variable names, why?

File: gym/spaces/graph.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+            num_edges = self.np_random.integers(num_nodes * num_nodes)
+
+        node_sample_space = self._generate_sample_space(self.node_space, num_nodes)
+        edge_sample_space = self._generate_sample_space(self.edge_space, num_edges)
+
+        sampled_nodes = self._sample_sample_space(node_sample_space)
+        sampled_edges = self._sample_sample_space(edge_sample_space)
+
+        sampled_node_list = None
+        if sampled_edges is not None:
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:174]: simplify the two if statement into one

File: gym/spaces/graph.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+from gym.spaces.discrete import Discrete
+from gym.spaces.multi_discrete import MultiDiscrete
+from gym.spaces.space import Space
+from gym.utils import seeding
+
+
+class GraphObj:
+    r"""A base for constructing information as graphs."""
+
+    def __init__(
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:16]: I haven't done much with graphs in a long time, but can they not be represented by just an adjacency matrix and a node list (I guess, though not necessary)? Have I missed something, but how can be a graph be represented by just a list of nodes?

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Im the most familiar with graphs for neural networks, is there a prior project you are basing this code off? Or that it would be easily compatible with? 
Sorry, if my questions don't make sense, I just want to understand, what is happening

File: gym/spaces/graph.py (hunk 1 above)
Comment (jjshoots) [reply gym/spaces/graph.py:22]: Sure

File: gym/spaces/graph.py (hunk 2 above)
Comment (jjshoots) [reply gym/spaces/graph.py:14]: Sure

File: gym/spaces/graph.py (hunk 3 above)
Comment (jjshoots) [reply gym/spaces/graph.py:106]: I'm not completely sure about, I kinda just used the Box space as a template for building thid space, but I'll have a look again

File: gym/spaces/graph.py (hunk 4 above)
Comment (jjshoots) [reply gym/spaces/graph.py:129]: Sure

File: gym/spaces/graph.py (hunk 5 above)
Comment (jjshoots) [reply gym/spaces/graph.py:180]: The idea for using tuples can work, but also makes things much harder to enforce and perform checks for. But if that is desired, I can change it to be so.

`sampled_node_list` and `edge_links` is an artifact from some refactoring, thanks for catching this.

File: gym/spaces/graph.py (hunk 6 above)
Comment (jjshoots) [reply gym/spaces/graph.py:174]: Sure

File: gym/spaces/graph.py (hunk 7 above)
Comment (jjshoots) [reply gym/spaces/graph.py:16]: Yes they can, but that comes with 2 immediate downsides - 1. n^2 instead of 2n memory complexity (where n can be m^2 for m nodes), 2. Inability to represent directional edges without having two adjacency matrices (or one with different tokens representing different connections, which further complicates things). Though it is a much rarer use case, an adjacency matrix also can't represent multiple edges between two nodes.

Using a list makes things much easier for computation using GNNs or transformer-like models as well, since most of the time the `edge_links` can just be concatenated on top of the `edges` and directly passed to the model.

In any case, one should be able to construct an adjacency matrix using `edge_links` fairly easily since it's already a list of indices.

Comment (jjshoots) [reply gym/spaces/graph.py:16]: In any case, if an adjacency matrix is the preferred way to do things, I propose doing so using a wrapper instead.

File: gym/spaces/graph.py (hunk 3 above)
Comment (jjshoots) [reply gym/spaces/graph.py:106]: It appears I stole template code from an older version of Box, I'll amend things.

File: gym/spaces/graph.py [hunk 8]
```diff
@@ -84,6 +19,8 @@ class Graph(Space):
         self.observation_space = spaces.Graph(node_space=space.Box(low=-100, high=100, shape=(3,)), edge_space=spaces.Discrete(3))
     """
 
+    _graph_obj_ctor = namedtuple("graph_obj", ["nodes", "edges", "edge_links"])
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:22]: Normally, NamedTuples are specified in global scope like a normal class. You could call it `GraphInstance`?

File: gym/spaces/graph.py [hunk 9]
```diff
@@ -114,12 +51,31 @@ def __init__(
...
-        # graph object creator
-        self.graphObj = GraphObj
-
         super().__init__(None, None, seed)
 
-    def _generate_sample_space(self, base_space: Union[None, Box, Discrete], num: int) -> Optional[Union[Box, Discrete]]:
+    @staticmethod
+    def graph_obj(
+        nodes: np.ndarray, edges: np.ndarray, edge_links: np.array
+    ) -> NamedTuple:
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:59]: With the named tuple in the global scope then I don't think this function is needed.

File: gym/spaces/graph.py [hunk 10]
```diff
@@ -147,8 +103,12 @@ def _sample_sample_space(self, sample_space) -> Optional[np.ndarray]:
         else:
             return None
 
-    def sample(self) -> GraphObj:
-        """Returns a random sized graph space with num_nodes between 1 and 10."""
+    def sample(self) -> NamedTuple:
+        """Generates a single sample graph with num_nodes between 1 and 10 sampled from the Graph.
+
+        Returns:
+            A NamedTuple representing a graph with attributes .nodes, .edges, and .edge_links.
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:110]: I think the attributes should be in quotes not as attributes but this doesn't matter much

File: gym/spaces/graph.py [hunk 11]
```diff
@@ -164,17 +124,18 @@ def sample(self) -> GraphObj:
...
+        if sampled_edges is not None and num_edges > 0:
+            sampled_edge_links = self.np_random.integers(
+                low=0, high=num_edges, size=(num_edges, 2)
+            )
 
-        return GraphObj(sampled_nodes, sampled_edges, sampled_edge_links)
+        return Graph.graph_obj(sampled_nodes, sampled_edges, sampled_edge_links)
 
-    def contains(self, x: GraphObj) -> bool:
+    def contains(self, x: NamedTuple) -> bool:
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:134]: The type can be the named tuple object

File: gym/spaces/graph.py [hunk 12]
```diff
@@ -164,17 +124,18 @@ def sample(self) -> GraphObj:
...
         """Return boolean specifying if x is a valid member of this space."""
-        if not isinstance(x, GraphObj):
+        if not isinstance(x, Graph._graph_obj_ctor):
+            print(type(x))
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:137]: I think this is left over from testing

File: gym/spaces/graph.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+            edge_space (Union[None, Box, Discrete]): space of the node features.
+            seed: Optionally, you can use this argument to seed the RNG that is used to sample from the space.
+        """
+        assert isinstance(
+            node_space, (Box, Discrete)
+        ), "Values of the node_space should be instances of Box or Discrete"
+        if edge_space is not None:
+            assert isinstance(
+                edge_space, (Box, Discrete)
+            ), "Values of the edge_space should be instances of Box or Discrete"
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:64]: A repeat of the debug comment above

File: gym/spaces/graph.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+            sampled_edge_links = self.np_random.integers(
+                low=0, high=num_edges, size=(num_edges, 2)
+            )
+
+        return GraphInstance(sampled_nodes, sampled_edges, sampled_edge_links)
+
+    def contains(self, x: GraphInstance) -> bool:
+        """Return boolean specifying if x is a valid member of this space."""
+        if not isinstance(x, GraphInstance):
+            print("fail 1")
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/graph.py:132]: Remove prints, repeat for all in `contains`

File: gym/spaces/utils.py [hunk 15]
```diff
@@ -118,6 +131,24 @@ def _flatten_dict(space, x) -> np.ndarray:
     return np.concatenate([flatten(s, x[key]) for key, s in space.spaces.items()])
 
 
+@flatten.register(Graph)
+def _flatten_graph(space, x) -> np.ndarray:
+    def _graph_unflatten(space, x):
+        ret = None
+        if space is not None and x is not None:
+            if isinstance(space, Box):
+                ret = x.reshape(x.shape[0], -1)
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/utils.py:140]: Can we call `flatten(x, space)` for both Box and Discrete here? Just in case the specifications of Box and Discrete change

File: gym/spaces/utils.py [hunk 16]
```diff
@@ -181,14 +212,34 @@ def _unflatten_dict(space: Dict, x: np.ndarray) -> dict:
     )
 
 
+@unflatten.register(Graph)
+def _unflatten_graph(space: Graph, x: GraphInstance) -> GraphInstance:
+    def _graph_unflatten(space, x):
+        ret = None
+        if space is not None and x is not None:
+            if isinstance(space, Box):
+                ret = x.reshape(-1, *space.shape)
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/utils.py:221]: Same comment just to call `unflatten(space, x)`

**Review by pseudo-rnd-thoughts — APPROVED** [review]
Overall looks good to merge with the comments addressed

File: gym/spaces/utils.py [hunk 17]
```diff
@@ -181,14 +219,40 @@ def _unflatten_dict(space: Dict, x: np.ndarray) -> dict:
     )
 
 
+@unflatten.register(Graph)
+def _unflatten_graph(space: Graph, x: GraphInstance) -> GraphInstance:
+    """We're not using `.unflatten()` here because a graph is not a homogeneous space.
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/utils.py:224]: "We're not using `.unflatten()` for :class:`Box` and :class:`Space`"

File: gym/spaces/utils.py [hunk 18]
```diff
@@ -118,6 +136,26 @@ def _flatten_dict(space, x) -> np.ndarray:
     return np.concatenate([flatten(s, x[key]) for key, s in space.spaces.items()])
 
 
+@flatten.register(Graph)
+def _flatten_graph(space, x) -> np.ndarray:
+    """We're not using `.flatten()` here because a graph is not a homogeneous space, see `.flatten` docstring."""
```
Comment (pseudo-rnd-thoughts) [inline gym/spaces/utils.py:141]: "We're not using `.flatten()` for :class:`Box` and :class:`Space`"
